lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -0,0 +1,1004 @@
+      updatePairByConnecting(ExCnt, E->ExecutionCount / To->OutSum, Dist, ExCnt,
+                             1.0, To->Size + E->DataSize);
+      if (ExCnt > 1e-3)
+        updateCPGBetweenNodes(From, E->ToNode, E, ExCnt, Dist, CheckList);
+    }
+  }
+  CheckList->erase(To);
+}
+
+// Main function to update a node to its subtree. It's a wrapper function to
+// call the updateCPGBetweenNodes() function recursively. The function is used
+// when disconnecting an edge in collapsing FRG
+void CloseProximityBuilder::updateCPGFromNodeToSubtree(
+    FieldReferenceGraph::Edge *E) {
+  if (E->FromNode->FieldNum == 0 || E->ExecutionCount <= 1e-3)
+    return;
+  DEBUG_WITH_TYPE(DEBUG_TYPE_CPG, dbgs() << "Update CPG from "
+                                         << E->FromNode->Id << " to subtree "
+                                         << E->ToNode->Id << "\n");
+  NodeSetType CheckList;
+  updateCPGBetweenNodes(E->FromNode, E->ToNode, E, E->ExecutionCount,
+                        E->DataSize, &CheckList);
+}
+
+// Core function to collapse FRG. Edge Arc is the edge that is going to be
+// collapsed, i.e. only has one predecessor. Collapse this Edge means to add
+// the direct successor of this Edge (Arc->ToNode) to the collapesd entries
+// of this edge. If the node has outgoing edges, which should be a common case,
+// there are two possible solutions depending on if the edge is collapsed:
+// 1. If the outgoing edge is collapsed, i.e. it connects to a transformed leaf
+//    node, we only need to update CP relations between the source Node and all
+//    the collapsed entries and also move those collapsed entries to the Edge
+//    Arc. Before update CP relations or move collapsed entries, we need to
+//    recalculate the count in the entry according to the Order 2 above (see
+//    comments before updateCPGBetweenNodes() ).
+// 2. If the outgoing edge is not collapsed, we need to move the subtree to the
+//    source Node (Arc->FromNode), because the node Arc->ToNode is going to be
+//    collapsed and "disappeared". This disconnection needs to call
+//    updateCPGFromNodeToSubtree() to make sure we don't lose any information
+//    and the reconnection needs to call Edge::reconnect() to update weights.
+bool CloseProximityBuilder::collapseSuccessor(FieldReferenceGraph *FRG,
+                                              FieldReferenceGraph::Edge *Arc) {
+  bool Ret = false;
+  DEBUG_WITH_TYPE(DEBUG_TYPE_CPG, dbgs() << "Collapse successor Edge ("
+                                         << Arc->FromNode->Id << ","
+                                         << Arc->ToNode->Id << ")\n");
+  assert(!std::isnan(Arc->ExecutionCount));
+  updateCPG(Arc->FromNode->FieldNum, Arc->ToNode->FieldNum, Arc->ExecutionCount,
+            Arc->DataSize);
+  FRG->collapseNodeToEdge(Arc->ToNode, Arc);
+  auto Edges = Arc->ToNode->OutEdges;
+  assert(Arc->ToNode->InSum > 0);
+  auto Ratio = Arc->ExecutionCount / Arc->ToNode->InSum;
```

Comment:
I'm not adept with C++11 & C++14, but...it looks to me like you are dividing an integer by an integer, so wouldn't 'auto' here decide the result type for Ratio is also an integer, and therefore instead of getting the fraction I think you want, it will be rounded to a 1 or a zero?  (If I'm wrong here, which is quite possible, I would love to have it explained to me.)